src/JUnit.Xml.TestLogger/JunitXmlSerializer.cs

```diff
@@ -0,0 +1,370 @@
+
+            return new TestSuite
+            {
+                Element = element,
+                Name = name,
+                FullName = fullName,
+                Total = total,
+                Passed = passed,
+                Failed = failed,
+                Inconclusive = inconclusive,
+                Skipped = skipped,
+                Error = error,
+                Time = time
+            };
+        }
+
+        /// <summary>
+        /// Produces a consistently indented output, taking into account that incoming messages
+        /// often have new lines within a message.
+        /// </summary>
+        private static string Indent(IReadOnlyCollection<TestResultMessage> messages)
+        {
+            var indent = "    ";
+
+            // Splitting on any line feed or carrage return because a message may include new lines
+            // that are inconsistent with the Environment.NewLine. We then remove all blank lines so
+            // it shouldn't cause an issue that this generates extra line breaks.
+            return
+                indent +
+                string.Join(
+                    $"{Environment.NewLine}{indent}",
+                    messages.SelectMany(m =>
+                        m.Text.Split(new string[] { "\r", "\n" }, StringSplitOptions.None)
+                              .Where(x => !string.IsNullOrWhiteSpace(x))
+                              .Select(x => x.Trim())));
+        }
+
+        private XElement CreateTestSuitesElement(
+            List<TestResultInfo> results,
+            TestRunConfiguration runConfiguration)
+        {
+            var assemblies = results.Select(x => x.AssemblyPath).Distinct().ToList();
+            var testsuiteElements = assemblies
+                .Select(a => this.CreateTestSuiteElement(
+                    results.Where(x => x.AssemblyPath == a).ToList(),
+                    runConfiguration));
+
+            return new XElement("testsuites", testsuiteElements);
+        }
+
+        private XElement CreateTestSuiteElement(List<TestResultInfo> results, TestRunConfiguration runConfiguration)
+        {
+            var testCaseElements = results.Select(a => this.CreateTestCaseElement(a));
+
+            StringBuilder stdOut = new StringBuilder();
+            foreach (var m in results.SelectMany(x => x.Messages))
+            {
+                if (TestResultMessage.StandardOutCategory.Equals(m.Category, StringComparison.OrdinalIgnoreCase))
+                {
+                    stdOut.AppendLine(m.Text);
+                }
+            }
+
+            StringBuilder stdErr = new StringBuilder();
+            foreach (var m in results.SelectMany(x => x.Messages))
+            {
+                if (TestResultMessage.StandardErrorCategory.Equals(m.Category, StringComparison.OrdinalIgnoreCase))
+                {
+                    stdOut.AppendLine(m.Text);
+                }
+            }
```

Comment:
@codito The acceptance test that checks for stdErr is failing. Any idea why this isn't working? The stdOut is passing based on the code right above it.

Reply:
@Siphonophora there may be a typo in line 209, `stdErr` instead of `stdOut`.

Reply:
@codito Oh yeah thanks for spotting that. Unfortunately that didn't fix it. Should the testlogger be capturing text from `Console.Error.WriteLine`?

Reply:
There may be a bit of confusion. The `stderr` messages are actually thrown by the Test Framework, not the individual tests, and those are captured separately in Test Logger (see `TestRunMessageWorkflow` in testlogger). These were earlier captured by the loggers directly: https://github.com/spekt/junit.testlogger/blob/144c2416ebecd0282d17785c7768bdaf767166b5/src/JUnit.Xml.TestLogger/JUnitXmlTestLogger.cs#L269

This is not same as the `TestResult.Messages`, which only appears to capture the stdout messages for a single TestResult. https://github.com/spekt/junit.testlogger/blob/144c2416ebecd0282d17785c7768bdaf767166b5/src/JUnit.Xml.TestLogger/JUnitXmlTestLogger.cs#L303

We will need to expose the run level messages to Serializer in this call: https://github.com/spekt/testlogger/blob/f5704c09657c020278c25f7d344f6648274530c8/src/TestLogger/Core/TestRunCompleteWorkflow.cs#L38.

Reply:
Please do let me know if you'd like to take a shot at fixing the Serializer API in testlogger

Reply:
Yeah, I think I understand what you mean. The Junit doesn't support run level test messages so I haven't been thinking in those terms much. 

I was looking at the code the other day and wasn't sure how the messages were getting into the TestResultInfo. I'm was kind of thrown by this, because it looked like something was supposed to happen with the messages in the `TestResultStore` but was missing. I saw the Xunit adapter isn't implemented yet, so maybe something special was going to happen there.
    https://github.com/spekt/testlogger/blob/f5704c09657c020278c25f7d344f6648274530c8/src/TestLogger/Extensions/DefaultTestAdapter.cs#L9-L16

For the change, are you just thinking we would add a parameter for `messages` here, or something more complex? I can take care of it either way.

https://github.com/spekt/testlogger/blob/f5704c09657c020278c25f7d344f6648274530c8/src/TestLogger/Core/TestRunCompleteWorkflow.cs#L38-L41

Reply:
> the Xunit adapter isn't implemented yet, so maybe something special was going to happen there.
https://github.com/spekt/testlogger/blob/f5704c09657c020278c25f7d344f6648274530c8/src/TestLogger/Extensions/DefaultTestAdapter.cs#L9-L16

This is an interesting case. Xunit test framework has a unique behavior to output the `skipped` test reasons as _run level_ messages. That forced us to create https://github.com/spekt/testlogger/blob/master/src/TestLogger/Extensions/XunitTestAdapter.cs#L19 so that the loggers do not replicate adapter specific logic. Fortunately, other test frameworks do not need this, hence `DefaultTestAdapter` remains empty :)

> For the change, are you just thinking we would add a parameter for messages here, or something more complex?

We can just add a new parameter there.